src/Promise.php

```diff
@@ -263,10 +263,17 @@ private function invokeWaitList()
         $this->waitList = null;
 
         foreach ($waitList as $result) {
-            $result->waitIfPending();
-            while ($result->result instanceof Promise) {
-                $result = $result->result;
+            while (true) {
                 $result->waitIfPending();
+
+                if ($result->result instanceof Promise) {
+                    $result = $result->result;
+                } else {
+                    if ($result->result instanceof PromiseInterface) {
+                        $result->result->wait(false);
```

Comment:
`Promise\invokeWaitFn()` doesn't throw on rejected promises, so `wait(false)` essentially invokes the same behaviour with custom implementations of `PromiseInterface`. Presumably, any errors we encounter in the wait list could be unrelated to `$this` and so shouldn't bubble up.

Furthermore, if we threw here, wouldn't that prevent the later promises in `$waitList` from being awaited, this preventing `$this` from being resolved/rejected?

I think the actual unwrapping happens in `wait()` once the wait list has been invoked.